File: leetcode/search_insert_position.erl
%% https://leetcode.com/problems/search-insert-position/
-module(search_insert_position).

-export([search_insert/2]).

-spec search_insert(Nums :: [integer()], Target :: integer()) -> integer().
search_insert(Nums, Target) -> search_insert(Nums, Target, 0, length(Nums), 0).

search_insert([First | _Others], Key, _Lower, _Upper, _Res) when First >= Key -> 0;
% search_insert([First | _Others], Key, _Lower, _Upper, _Res) when First >= Key -> 0;
search_insert(_, _Key, Lower, Upper, Res) when Lower > Upper -> Res;
search_insert(A, Key, Lower, Upper, Res) ->
    Mid = Lower + (Upper - Lower) div 2,
    Item = lists:nth(Mid+1, A),
    Last = lists:last(A),
    if
        Last < Key -> Upper;
        Key == Item -> Mid;
        Key < Item -> search_insert(A, Key, Lower, Mid - 1, Mid);
        Key > Item -> search_insert(A, Key, Mid + 1, Upper, Res)
    end.

File: leetcode/_Other/search_insert_position.c
// https://leetcode.com/problems/search-insert-position/
#include<stdlib.h>
#include<string.h>
#include<stdio.h>

#define DEBUG

#ifdef DEBUG
int max_it = 0;
#endif

int searchInsert(int* nums, int numsSize, int target) {
    
    if (nums[0] >= target) {
        return 0;
    }
    if (nums[numsSize-1] < target) {
        return numsSize;
    }

    int low = 0, high = numsSize, result = 0;

    while (low <= high) {
        int mid = low + (high - low) / 2;

        if (nums[mid] == target) {
            return mid;
        }
        if (target < nums[mid]) {
            result = mid;
            high = mid - 1;
        }
        else {
            result = mid + 1;
            low = mid + 1;
        }

    }


    return result;
    // int inx = numsSize / 2;


        // int mid;
        // while (low <= high) {
        //     mid = low + (high - low) / 2;
        //     if (nums[mid] <= target) {
        //         result = mid;
        //         low = mid + 1;
        //     }
        //     else {
        //         high = mid - 1;
        //     }
        // }
    // return result;
// 
}

int main() {
    int nums[4] = {1,3,5,6};
    printf("%d\n", searchInsert(nums, 4, 7));
}
